grpc-client-server/prober_client/main.go

```diff
@@ -26,13 +62,73 @@ func main() {
 	c := pb.NewProberClient(conn)
 
 	// Contact the server and print out its response.
-	ctx := context.Background() // TODO: add a timeout
+	var wg sync.WaitGroup
+	for _, endpoint := range endpoints {
+		req := &pb.ProbeRequest{Endpoint: endpoint, NumberOfRequests: int32(*nOfRequests)}
+		wg.Add(1)
+		go SingleProbe(os.Stdout, c, req, &wg)
+	}
+	wg.Wait()
+}
+
+func SingleProbe(w io.Writer, c pb.ProberClient, req *pb.ProbeRequest, wg *sync.WaitGroup) {
+	results := make(chan *Result)
+	timeout := time.Duration(*timeout) * time.Second
+	go CreateProgressBar(w, timeout, req.Endpoint, results, wg)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	r, err := c.DoProbes(ctx, req)
+	cancel()
```

Comment:
Why are you calling cancel here? What does that do?

Reply:
From my understanding, we call cancel to release resources if we are done with the context. In this case the request has completed (error or not), so there's no need to keep the context and we can terminate it?